### PR TITLE
Fixed Some bugs of code_mobject.py and Text_mobject.py

### PR DESCRIPTION
**Pull request by NavpreetDevpuri**
Reopening https://github.com/ManimCommunity/manim/pull/90
Fixed https://github.com/3b1b/manim/issues/1067
NOTE : SurroundingRectangle() only contains visible text.
```Python
class te(Scene):
    def construct(self):
        text1 = Text(" ab ", font="Consolas", size=2)
        rect1 = SurroundingRectangle(text1)
        text1.chars[1].set_color(GREEN)
        self.add(text1,rect1)
```
Output
te

```python
class bug1(Scene):
    def construct(self):
        text1 = Text("  ab\ncd", font="Consolas", size=2)
        text2 = Text("ab\ngh", font="Consolas", size=2)
        rect1 = SurroundingRectangle(text1)
        rect2 = SurroundingRectangle(text2)
        self.play(Transform(remove_invisible_chars(text1), remove_invisible_chars(text2)), Transform(rect1, rect2))
        self.wait()
```
Output 
bug1

Added new parameters `background_stroke_width` and `background_stroke_color`
```python
class bug2andbug3(Scene):
    def construct(self):
        helloworldc = Code(
            "helloworldc.c",
            tab_width=4,
            background_stroke_width=1,
            background_stroke_color=WHITE,
            insert_line_no=True,
            style=Code.styles_list[4],
            background="window",
            language="cpp",
        )
        helloworldcpp = Code(
            "helloworldcpp.cpp",
            tab_width=4,
            background_stroke_width=1,
            background_stroke_color=WHITE,
            insert_line_no=True,
            style=Code.styles_list[15],
            background="window",
            language="cpp",
        )
        helloworldc.move_to(np.array([-3.6, 0, 0]))
        helloworldcpp.move_to(np.array([3.1, 0, 0]))
        self.add(helloworldc,helloworldcpp)
```
Output
bug2andbug3

Updated Paragraph() and added new methods `set_line_to_initial_position()` and `set_all_lines_to_initial_positions()`
```python
class para(Scene):
    def construct(self):
        t = Paragraph('this is a awesome', 'paragraph', 'With \nNewlines', '\tWith Tabs', '  With Spaces',
                      'With Alignments',
                      'center', "left", "right")
        t.set_line_alignment("center", 7)
        t.set_line_alignment("right", 9)
        t.shift(2 * RIGHT)
        rect = SurroundingRectangle(t)
        self.add(t, rect)
        self.wait()
        self.play(t.set_all_lines_alignments, "right")
        self.play(t.set_line_to_initial_position, 4)
        self.play(t.set_all_lines_to_initial_positions)
        self.play(t.set_line_alignment, "center", 7)
        self.play(t.set_line_alignment, "right", 9)
        t.chars[0][5:7].set_color(GREEN)
        t[1][0:4].set_color(YELLOW)
        self.wait()
```
Output
ezgif-3-3dc23c79c5fa

### CODE REVIEW

**Comment by NavpreetDevpuri** [comment]
@Aathish04 thanks for that. i did not notice that.
Now i fixed it. check it.

**Comment by huguesdevimeux** [comment]
I didn’t follow this PR at all. Is it ready ?

**Comment by NavpreetDevpuri** [comment]
> I didn’t follow this PR at all. Is it ready ?

I think it will work. 
But i didn't tested it by myself. I installed new copy of windows and manim's development environment is deleted. I will test it myself.

**Comment by Aathish04** [comment]
> > I didn’t follow this PR at all. Is it ready ?
> 
> I think it will work.
> But i didn't tested it by myself. I installed new copy of windows and manim's development environment is deleted. I will test it myself.

@NavpreetDevpuri I'm so sorry for taking way too much time to respond. There are two things that I think need to be done before this is ready to merge:

a) Format the code via `black`,
b) Fix merge conflicts between your branch and master.

Once these are resolved, I think we'll be ready to merge.

**Comment by NavpreetDevpuri** [comment]
I tested it and it works fine for me.
Here are some outputs
## Code
```python
class Test6(Scene):
    def construct(self):
        t = Paragraph('this is a awesome', 'paragraph', 'With \nNewlines', '\tWith Tabs', '  With Spaces',
                      'With Alignments',
                      'center', "left", "right", line_spacing=0.3, size=1, alignment="left", t2c={"Tabs": RED})
        # t.set_alignment("center", 7)
        #self.play(t.shift, RIGHT)
        t.__len__()
        self.play(t.set_color, RED)
        t.set_line_alignment("left", 8)
        t.set_line_alignment("right", 9)
        #self.play(t.set_line_alignment, "right", 7)
        self.play(t.set_all_lines_alignments, "right")
        # t[0][5].set_color(GREEN)
        # t[0][6].set_color(GREEN)
        # t[1][0:4].set_color(YELLOW)
        rect = SurroundingRectangle(t)
        self.add(t, rect)
        self.wait()
```
## Output
Test6
<br>
## Code
```python
class test15(Scene):
    def construct(self):
        helloworldc = Code(
            "helloworldc.c",
            tab_width=4,
            background_stroke_width=1,
            background_stroke_color=WHITE,
            insert_line_no=True,
            style=Code.styles_list[4],
            background="window",
        )
        helloworldcpp = Code(
            "helloworldcpp.cpp",
            tab_width=4,
            background_stroke_width=1,
            background_stroke_color=WHITE,
            insert_line_no=True,
            style=Code.styles_list[15],
            background="window",
        )
        #helloworldc.move_to(np.array([-3.6, 0, 0]))
        #helloworldcpp.move_to(np.array([3.1, 0, 0]))
        self.add(helloworldc)
        self.play(Transform(remove_invisible_chars(helloworldc),
                            remove_invisible_chars(helloworldcpp)))
        # helloworldc[2][0:4].set_color(RED)
        # self.play(Transform(helloworldc[0], helloworldcpp[0]),
        #Transform(helloworldc[1], helloworldcpp[1]),
        # Transform(helloworldc[2], helloworldcpp[2]))
        self.wait()
```

## Output 
test15

**Comment by NavpreetDevpuri** [comment]
Should i check for file in `assets\codes\` ? Or should i just need to ask user to specify a direct link to file ?

Code_YFF2g0itOV

**Comment by NavpreetDevpuri** [comment]
You guys removed `conts` from `Constants.py` !
Let me correct `TextMobject.py`
Code_gOda9Moy1j

**Comment by Aathish04** [comment]
> I am correcting some bugs so that it also works for manim-community.
> In `CodeMobject.py` file, Should i check for file in `assets\codes\` ? Or should i just need to ask user to specify a direct link to file ?

What asset is `Code()` looking for? In general, I think it's better to use the `assets` directory like you have done here, but others may have some other thoughts...

**Comment by NavpreetDevpuri** [comment]
> What asset is `Code()` looking for? In general, I think it's better to use the `assets` directory like you have done here, but others may have some other thoughts...

Its looking for specified file `helloworldc = Code("helloworldc.c",...etc`. The way i did that so, we can specify a direct link to file or just a name of file that we already copied in `assets\codes`

**Comment by NavpreetDevpuri** [comment]
> You guys removed `conts` from `Constants.py` !
> Let me correct `text_mobject.py`
> Code_gOda9Moy1j




### Why is this happening ? Anyone ?

**Comment by NavpreetDevpuri** [comment]
Now tested on manim-community and it works fine 😃

**Comment by leotrs** [comment]
@NavpreetDevpuri thank you for the update. The tests are failing. Could you please check to see what's the problem?

**Comment by NavpreetDevpuri** [comment]
Who, how, and why someone added this checks for code formates. I am using prettifier in visual studio code and it puts `a:5` as `a: 5` but checker want it to be `a : 5`, I think that's wrong and it will slow development process. Remove those checks or make them loose, so that it can accept both formats. 

chrome_Pvt9tHDemk

**Comment by leotrs** [comment]
> Who, how, and why someone added this checks for code formates.

Who: the core dev team. How: by using GH actions and `black`. Why: you can read all about it in #172, #174, #43, #197, etc.

> I am using prettifier in visual studio code and it puts `a:5` as `a: 5` but checker want it to be `a : 5` instead of `a: 5`

I suggest you use a tool that is compatible with ManimCommunity's code standards.

> I think that's wrong 

There are no rights or wrongs when it comes to code formatting. There are things you like and things you don't like.

> and it will slow development process.

It will also standardize code and make the entire codebase more readable and easier to debug.

> Remove those checks or make them loose, so that it can accept both formats.

No. If you have a problem with this repository's coding standards, please open an issue to discuss any desired changes and let everyone else chime in. The community's standards will not be changed only so that your code prettifier doesn't like our standards.

**Comment by NavpreetDevpuri** [comment]
I don't understand these `Checks`. Can anyone please help me with this?

**Comment by kilacoda** [comment]
> I don't understand these `Checks`. Can anyone please help me with this?

They're tests meant to see if your changes don't break any functionality of the library.

cc'ing @huguesdevimeux to explain what the errors and fails mean.

**Comment by NavpreetDevpuri** [comment]
Yes, this code is changing the basic nature of `Text()` as it is fixing the bug in `Text()` as mentioned in a failed test. 
But I am not sure about it.

chrome_6Ph1ixNvIt

**Comment by huguesdevimeux** [comment]
This is weird. Tests are working fine on my computer

File: manim/mobject/svg/text_mobject.py
Comment (Aathish04) [inline manim/mobject/svg/text_mobject.py:430]: Why does this run `Container.__init__` if it inherits from VGroup?

File: manim/mobject/svg/code_mobject.py
Comment (Aathish04) [inline manim/mobject/svg/code_mobject.py:4]:
```suggestion
```
Is this actually necessary? You run `Container.__init__` for `Paragraph`, but `Paragraph` inherits from `VGroup` not `Container`.

**Review by Aathish04 — DISMISSED** [review]
@NavpreetDevpuri 
I've requested some changes.

Also, did you check whether the tests passed on your new computer? They still seem to fail on mine, and on the Github Actions runner.

@huguesdevimeux Can you please check again if the tests pass or fail on your system?

I actually think the test reference data may need to be remade for this PR, similar to what happened with @philippimhof 's changes.

The _moment_ all tests pass, this is being merged into Master.

File: manim/mobject/svg/text_mobject.py
Comment (NavpreetDevpuri) [reply manim/mobject/svg/text_mobject.py:430]: I think i did it just to `digest_config(self, config)`
Let me try this in that way.

Comment (NavpreetDevpuri) [reply manim/mobject/svg/text_mobject.py:430]: i needed init because i thought i need to init according to our need you can check. I mean custom init.

Comment (Aathish04) [reply manim/mobject/svg/text_mobject.py:430]: Yes, but why are you running _`Container`s_ `__init__` method? Shouldn't you run `VGroup.__init__` since `Paragraph` inherits from `VGroup`?

**Comment by NavpreetDevpuri** [comment]
Why? Restyling found differences?

I used black to format!

**Comment by leotrs** [comment]
Please ignore restyled for now. We just set it up and we are just trying it
out. Thanks for looking into this PR again :)

**Comment by NavpreetDevpuri** [comment]
> Please ignore restyled for now. We just set it up and we are just trying it out. Thanks for looking into this PR again :)

I copied code from the restyled bot and now it's working!
let me know if I should use black?

**Comment by huguesdevimeux** [comment]
> Also, did you check whether the tests passed on your new computer? They still seem to fail on mine, and on the Github Actions runner.
> 
> @huguesdevimeux Can you please check again if the tests pass or fail on your system?

I don't think tests will pass. If they don't pass on your system as well, maybe @NavpreetDevpuri you should regenerate new data for the failing test.

**Comment by NavpreetDevpuri** [comment]
> Yes, this code is changing the basic nature of `Text()` as it is fixing the bug in `Text()` as mentioned in a failed test.
> But I am not sure about it. But I think this can be ignored as mentioned in Test line number 57.
> 
> chrome_6Ph1ixNvIt

This is the first time I encounter these checks and I have almost no I idea how to resolve but I think this PR really solves essential bugs in `TextMobject` so maybe that's why the checks are failed.

**Comment by huguesdevimeux** [comment]
You basically have to re-grenerate some reference data of the tests. See the wiki. 

Are tests working on your computer ?

**Comment by NavpreetDevpuri** [comment]
> You basically have to re-generate some reference data of the tests. See the wiki.
> 
> Are tests working on your computer?

For now, I have no idea how to run these checks on the local PC. But I tried some code that gives expected outputs and mentioned above in this PR.

**Comment by NavpreetDevpuri** [comment]
> You basically have to re-generate some reference data of the tests. See the wiki.

I will try.

**Comment by huguesdevimeux** [comment]
@NavpreetDevpuri Hi, I merged from master (and fixed the tests). 

But I just noticed that there are some code commented out and some docstrings that are not in the right format. Could you fix that ? It should be very quick.

File: manim/mobject/svg/text_mobject.py
Comment (huguesdevimeux) [inline manim/mobject/svg/text_mobject.py:388]: Why is it commented out?

File: manim/mobject/svg/text_mobject.py
Comment (huguesdevimeux) [inline manim/mobject/svg/text_mobject.py:413]: Question : I don't think this is in the right format. What do you think @PgBiel ?

File: manim/mobject/svg/text_mobject.py
Comment (huguesdevimeux) [inline manim/mobject/svg/text_mobject.py:395]: **Issue :** I think  this print statement should not be there.
```suggestion
```

**Review by huguesdevimeux — CHANGES_REQUESTED** [review]

File: manim/mobject/svg/text_mobject.py
Comment (NavpreetDevpuri) [reply manim/mobject/svg/text_mobject.py:388]: I don't exactly remember but I remember maybe it's from a very old version when I was trying to solve spaces problem in `Text`

Comment (NavpreetDevpuri) [reply manim/mobject/svg/text_mobject.py]: we can ignore it

File: manim/mobject/svg/text_mobject.py
Comment (NavpreetDevpuri) [reply manim/mobject/svg/text_mobject.py:406]: my English is weak 😅

Comment (huguesdevimeux) [reply manim/mobject/svg/text_mobject.py:406]: No ahah your english is not the issue. The issue is like I don't think this fits our guidelines for docstring.

**Comment by NavpreetDevpuri** [comment]
I think we should merge it somehow for now. 
because it shows style error because of other files
I will reformate all files then I will open a different PR for it.

**Comment by NavpreetDevpuri** [comment]
I reformated all files locally and when I run comment `git status` it shows no differences!

**Comment by NavpreetDevpuri** [comment]
I tried to reformate this file with `black` in visual studio and it does not change according to what these checks suggesting 
chrome_Xckxq7IBg1

**Comment by leotrs** [comment]
> I tried to reformate this file with `black` in visual studio and it does not change according to what these checks suggesting

Hi @NavpreetDevpuri thanks for looking into this again :) Have you made sure to update your version of black? It recently changed from version 19 to version 20. Also, don't worry too much about the black check, it has been failing for a few days now. 

If you do find you have to update your black version PLEASE do not run black on the whole codebase and push those changes on this PR. If necessary, run black only on the files you are editing on this PR.

@Aathish04 @huguesdevimeux please merge this even without black check passing :)

**Comment by huguesdevimeux** [comment]
Yes, black tests failing are ok for now as this will be changed soon. 

Let me know when you think you're done with this PR. We are just waiting @PgBiel aproval, and then we can merge !

**Comment by NavpreetDevpuri** [comment]
I did my best. I checked/tested/rechecked/tested until I feel `That's almost perfect`

**Comment by Aathish04** [comment]
Marking as ready for review!

File: manim/mobject/svg/code_mobject.py
Comment (PgBiel) [inline manim/mobject/svg/code_mobject.py:419]: Add a Returns section.

File: manim/mobject/svg/code_mobject.py
Comment (PgBiel) [inline manim/mobject/svg/code_mobject.py:325]: Returns?

File: manim/mobject/svg/text_mobject.py
Comment (PgBiel) [inline manim/mobject/svg/text_mobject.py:22]: Docstring

File: manim/mobject/svg/code_mobject.py
Comment (PgBiel) [inline manim/mobject/svg/code_mobject.py:42]: Shouldn't this be in `Code`'s documentation?

File: manim/mobject/svg/text_mobject.py
Comment (PgBiel) [inline manim/mobject/svg/text_mobject.py:54]: Shouldn't this be in `Text`'s Docstring?

File: manim/mobject/svg/text_mobject.py
Comment (PgBiel) [inline manim/mobject/svg/text_mobject.py:135]: What's this? A commented out function?

File: manim/mobject/svg/text_mobject.py
Comment (PgBiel) [inline manim/mobject/svg/text_mobject.py:330]: Should still have docstrings..

File: manim/mobject/svg/text_mobject.py
Comment (PgBiel) [reply manim/mobject/svg/text_mobject.py:406]: Yeah this should be in the class's docstring, while attrs should be in the Attributes section. Please apply this to other classes as well

File: manim/mobject/svg/text_mobject.py
Comment (PgBiel) [inline manim/mobject/svg/text_mobject.py:412]: This would be the Examples section

File: manim/mobject/svg/text_mobject.py
Comment (PgBiel) [inline manim/mobject/svg/text_mobject.py:485]: Docstring

File: manim/mobject/svg/code_mobject.py
Comment (PgBiel) [inline manim/mobject/svg/code_mobject.py:500]: Line between description and section
```suggestion

    Parameters
```

File: manim/mobject/svg/code_mobject.py
Comment (PgBiel) [inline manim/mobject/svg/code_mobject.py:504]: Returns

File: manim/mobject/svg/code_mobject.py
Comment (PgBiel) [inline manim/mobject/svg/code_mobject.py:499]:
```suggestion 
    """Function that inserts line numbers in the highlighted HTML code.
```

File: manim/mobject/svg/code_mobject.py
Comment (PgBiel) [inline manim/mobject/svg/code_mobject.py:498]: Shouldn't the function name include "html"?

File: manim/mobject/svg/text_mobject.py
Comment (PgBiel) [inline manim/mobject/svg/text_mobject.py:59]: Examples section etc

File: manim/mobject/svg/text_mobject.py
Comment (PgBiel) [inline manim/mobject/svg/text_mobject.py:155]: docstring

File: manim/mobject/svg/text_mobject.py
Comment (PgBiel) [inline manim/mobject/svg/text_mobject.py:184]: Docstring

**Review by PgBiel — CHANGES_REQUESTED** [review]
There's still a long way to go regarding docs. Please read the guidelines in our Wiki.

**Comment by leotrs** [comment]
This has been open for 2 months and the only thing that seems to be blocking the final merge is documentation. @NavpreetDevpuri do you think you have time to work on the documentation suggestions left by @PgBiel ?

Otherwise, we should just merge as is and worry about documenting this module later.

**Comment by NavpreetDevpuri** [comment]
> @NavpreetDevpuri do you think you have time to work on the documentation suggestions left by @PgBiel ?
I was busy with something. I am a little weak in the documentation, I mean it takes more time for me to do simple documentations

let me try.

File: manim/mobject/svg/code_mobject.py
Comment (NavpreetDevpuri) [reply manim/mobject/svg/code_mobject.py]: now I added it to the documentation but I did not, because this seems like Extra?

File: manim/mobject/svg/text_mobject.py
Comment (NavpreetDevpuri) [reply manim/mobject/svg/text_mobject.py:67]: I did not add docstring for Text because I don't know exactly for what some parameters are used and I also don't able to understand some methods deeply.

**Review by NavpreetDevpuri — COMMENTED** [review]
> This has been open for 2 months and the only thing that seems to be blocking the final merge is documentation. @NavpreetDevpuri do you think you have time to work on the documentation suggestions left by @PgBiel ?
> 
> Otherwise, we should just merge as-is and worry about documenting this module later.


I did almost what I can do. so, we should merge it after a little review. we can improve documentation in other PR. this PR created mainly for fixing very big bugs.

File: manim/mobject/svg/code_mobject.py
Comment (leotrs) [reply manim/mobject/svg/code_mobject.py]: See [here](https://docs.python.org/3/tutorial/modules.html#importing-from-a-package)

File: manim/mobject/svg/code_mobject.py
Comment (leotrs) [reply manim/mobject/svg/code_mobject.py:314]: If it makes sense for this function to not have a return statement, then let's leave it as is for now.

File: manim/mobject/svg/text_mobject.py
Comment (leotrs) [reply manim/mobject/svg/text_mobject.py:67]: Let's work on this on another PR

File: manim/mobject/svg/text_mobject.py
Comment (leotrs) [reply manim/mobject/svg/text_mobject.py:72]: Let's work on this on another PR

File: manim/mobject/svg/text_mobject.py
Comment (leotrs) [reply manim/mobject/svg/text_mobject.py:148]: @NavpreetDevpuri let's figure this one out and leave the rest of the documentation for a future PR.

**Comment by leotrs** [comment]
Thanks @NavpreetDevpuri for doing what you can about the documentation. There is just one instance that I would like to have sorted out and then we can merge this. See my comments above.

File: manim/mobject/svg/text_mobject.py
Comment (NavpreetDevpuri) [reply manim/mobject/svg/text_mobject.py]: very old function from an old implementation just removed it.

File: manim/mobject/svg/text_mobject.py
Comment (NavpreetDevpuri) [reply manim/mobject/svg/text_mobject.py:321]: I will maybe create some different `Mobject` only for backgrounds of other `Mobjects`

**Comment by leotrs** [comment]
I see you removed that last thing - is there anything else you wish to do before we merge this?

**Comment by NavpreetDevpuri** [comment]
@leotrs sir, all clear for now.

**Pull request merged**
